File: solutions/cpp/0414.cpp
class Solution {
 public:
  int thirdMax(vector<int>& nums) {
    long maxi = LONG_MIN;
    long secondMax = LONG_MIN;
    long thirdMax = LONG_MIN;

    for (const int num : nums)
      if (num > maxi) {
        thirdMax = secondMax;
        secondMax = maxi;
        maxi = num;
      } else if (maxi > num && num > secondMax) {
        thirdMax = secondMax;
        secondMax = num;
      } else if (secondMax > num && num > thirdMax) {
        thirdMax = num;
      }

    return thirdMax == LONG_MIN ? maxi : thirdMax;
  }
};

File: solutions/java/0414.java
public class Solution {
  public int thirdMax(int[] nums) {
    Queue<Integer> pq = new PriorityQueue<>();
    Set<Integer> seen = new HashSet<>();

    for (int num : nums)
      if (seen.add(num)) {
        pq.add(num);
        if (pq.size() > 3)
          pq.poll();
      }

    if (pq.size() == 2)
      pq.poll();

    return pq.peek();
  }
}